### DISCUSSION TITLE
What difference between GraphQLContext and DataFetchingEnvironment?

### DISCUSSION THREAD

**Discussion by walfud** [ExpediaGroup/graphql-kotlin#1137] [Q&A] [answered]
Docs say "GraphQL context based on the incoming request and make it available during execution", while DataFetchingEnvironment "This argument will be automatically populated and injected during the query execution but will not be included in the schema definition".
Both object is available during a whole fetch process, what difference between the two?

**Comment by dariuszkuc** [comment, marked as answer]
`DataFetchingEnvironment` contains `GraphQLContext` (which can be any object) as well as some additional information about the currently executed request (i.e. which field is being requested, what arguments were specified, etc).

**Reply by smyrick** [reply]
To add on to that answer `DataFetchingEnvironment` is provided by graphql-java and contains defined info about the current operation. It can provide the current field selection set, arguments provided, parent field info, etc.

The `GraphQLContext` is an object that you define for your server and can contain any number of properties. This is where you can add custom info at the start of the request. A common example would be parsing the HTTP headers and adding some info or token to be read later.

**Reply by walfud** [reply]
Very clear. Thanks
